GTG/gtk/browser/browser.py

```diff
@@ -420,8 +425,76 @@ def _init_accelerators(self):
                                        Gtk.AccelFlags.VISIBLE)
 
 # HELPER FUNCTIONS ##########################################################
-    def temporary_search(self, widget):
-        self.quickadd_entry.grab_focus()
+
+    def on_search_toggled(self, widget):
+        search = self.builder.get_object("search")
+        self.bar = self.builder.get_object("search_bar")
+        self.entry = self.builder.get_object("search_entry")
+        if self.bar.get_search_mode():
+            search.set_active(False)
+            self.bar.set_search_mode(False)
+        else:
+            search.set_active(True)
+            self.bar.set_search_mode(True)
+            self.entry.grab_focus()
+
+    def on_search(self, key, widget):
+        query = self.search_entry.get_text()
+
+        try:
+            parsed_query = parse_search_query(query)
+            self.apply_filter_on_panes(SEARCH_TAG, parameters=parsed_query,
+                                       refresh=True)
+        except InvalidQuery:
+            pass
+
+    def on_save_search(self, widget):
+        query = self.search_entry.get_text()
+        # ! at the beginning is reserved keyword for liblarch
+        if query.startswith('!'):
+            label = '_' + query
+        else:
+            label = query
+
+        # find possible name collisions
+        name, number = label, 1
+        already_search = False
+        while True:
+            tag = self.req.get_tag(name)
+            if tag is None:
+                break
+
+            if tag.is_search_tag() and tag.get_attribute("query") == query:
+                already_search = True
+                break
+
+            # this name is used, adding number
+            number += 1
+            name = label + ' ' + str(number)
+
+        if not already_search:
+            tag = self.req.new_search_tag(name, query)
+
+        # Apply new search right now
+        if self.tagtreeview is not None:
+            # Select new search in tagsidebar and apply it
+
+            # Make sure search tag parent is expanded
+            # (otherwise selection does not work)
+            self.expand_search_tag()
+
+            # Get iterator for new search tag
+            model = self.tagtreeview.get_model()
+            path = self.tagtree.get_paths_for_node(tag.get_id())[0]
+            tag_iter = model.my_get_iter(path)
+
+            # Select only it and apply filters on top of that
+            selection = self.tagtreeview.get_selection()
+            selection.unselect_all()
+            selection.select_iter(tag_iter)
+            self.on_select_tag()
```

Comment:
Can you refactor this logic into a separate method, let's say `select_search_tag(tag_id)` ?